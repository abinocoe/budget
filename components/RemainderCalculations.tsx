import React, { Component } from "react";
import { StyleSheet, Text, View } from "react-native";
import Modal from "react-native-modal";

import * as colours from "../constants/colours";

import DateSelection from "./DateSelection";
import PeriodAllowanceInput from "./PeriodAllowance";

interface Props {
  dailyAllowance: number;
  daysElapsed: number;
  daysRemaining: number;
  intervalStartDate: number;
  totalAllowance: number;
  totalSpent: number;
  updateAllowance: (amount: number) => void;
  updateIntervalStartDate: (intervalStart: number) => void;
}

interface State {
  showModal: boolean;
  modalContent: string;
}

class RemainderCalculations extends Component<Props, State> {
  public constructor(props: Props) {
    super(props);

    this.state = {
      showModal: false,
      modalContent: ""
    };
  }

  public showModal = (contentString: string) => {
    this.setState({ showModal: true, modalContent: contentString });
  };

  public hideModal = () => {
    this.setState({ showModal: false, modalContent: "" });
  };

  public render() {
    const {
      dailyAllowance,
      daysElapsed,
      daysRemaining,
      intervalStartDate,
      totalAllowance,
      totalSpent,
      updateAllowance,
      updateIntervalStartDate
    } = this.props;
    const amountRemaining = totalAllowance - totalSpent;
    const decimalise = (amount: number) => (amount / 100).toFixed(2);

    const creditOrDebitAmount =
      // daysRemaining includes an extra day to always include the current day
      (totalAllowance / (daysRemaining - 1 + daysElapsed)) * daysElapsed -
      totalSpent;
    return (
      <View
        style={{
          marginTop: 20,
          flexDirection: "row",
          justifyContent: "space-evenly"
        }}
      >
        <View style={styles.group}>
          <PeriodAllowanceInput
            allowanceAmount={totalAllowance}
            updateAllowance={updateAllowance}
          />
          <View style={{ alignItems: "center" }}>
            <Text
              onPress={() =>
                this.showModal(
                  "Your total allowance for this month minus the total you've spent so far"
                )
              }
              style={[styles.largeText, { backgroundColor: colours.purple }]}
              testID="totalRemain"
            >
              {decimalise(amountRemaining)}
            </Text>
            <Text style={styles.smallText}>Total remaining</Text>
          </View>
          <DateSelection
            intervalStartDate={intervalStartDate}
            updateIntervalStartDate={updateIntervalStartDate}
          />
        </View>
        <View style={styles.group}>
          <View style={{ alignItems: "center" }}>
            <Text
              onPress={() =>
                this.showModal(
                  "Your total allowance divided by the number of days in the start month"
                )
              }
              style={[styles.largeText, { backgroundColor: colours.green }]}
            >
              {decimalise(dailyAllowance)}
            </Text>
            <Text style={styles.smallText}>Daily allowance</Text>
          </View>
          <View style={{ alignItems: "center" }}>
            <Text
              onPress={() =>
                this.showModal(
                  "The total amount remaining divided by the number of days remaining (including today)"
                )
              }
              style={[styles.largeText, { backgroundColor: colours.yellow }]}
              testID="averageRemain"
            >
              {decimalise(amountRemaining / daysRemaining)}
            </Text>
            <Text style={styles.smallText}>Average per day remain</Text>
          </View>
          <View style={{ alignItems: "center" }}>
            <Text
              onPress={() =>
                this.showModal(
                  "Your total allowance for the days elapsed so far, minus the total spent"
                )
              }
              style={[styles.largeText, { backgroundColor: colours.orange }]}
              testID="amountDebit"
            >
              {decimalise(
                Math.abs(parseFloat((creditOrDebitAmount / 100).toFixed(2))) ===
                  0
                  ? Math.abs(creditOrDebitAmount)
                  : creditOrDebitAmount
              )}
            </Text>
            <Text style={styles.smallText}>
              {parseFloat(decimalise(creditOrDebitAmount)) >= -0
                ? `Amount credit`
                : `Amount debit`}
            </Text>
          </View>
        </View>
        <Modal
          isVisible={this.state.showModal}
          onBackButtonPress={this.hideModal}
          onBackdropPress={this.hideModal}
        >
          <View
            style={{
              backgroundColor: "white",
              justifyContent: "center",
              alignItems: "center"
            }}
          >
            <Text>{this.state.modalContent}</Text>
          </View>
        </Modal>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  largeText: {
    fontSize: 30,
    paddingHorizontal: 5,
    marginTop: 10,
    marginBottom: 5,
    borderRadius: 5,
    width: "90%",
    textAlign: "center",
    color: "white",
    elevation: 3
  },
  smallText: {
    fontSize: 12
  },
  group: {
    justifyContent: "flex-start"
  }
});

export default RemainderCalculations;
