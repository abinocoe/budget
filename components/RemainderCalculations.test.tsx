import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import RemainderCalculations from "./RemainderCalculations";

describe("RemainderCalculations", () => {
  it("renders correctly", () => {
    const output = shallow(
      <RemainderCalculations
        dailyAllowance={1000}
        daysElapsed={2}
        daysRemaining={29}
        intervalStartDate={25}
        totalAllowance={30000}
        totalSpent={3000}
        updateAllowance={() => null}
        updateIntervalStartDate={() => null}
      />
    );
    expect(toJson(output)).toMatchSnapshot();
  });

  it("calculates the total remaining", () => {
    const output = shallow(
      <RemainderCalculations
        dailyAllowance={1000}
        daysElapsed={2}
        daysRemaining={29}
        intervalStartDate={25}
        totalAllowance={30000}
        totalSpent={3000}
        updateAllowance={() => null}
        updateIntervalStartDate={() => null}
      />
    );
    const debitAmount = output.findWhere(
      node => node.prop("testID") === "amountDebit"
    );
    expect(debitAmount.first().prop("children")).toEqual("-10.00");
  });

  it("calculates the average left for remaining days", () => {
    const output = shallow(
      <RemainderCalculations
        dailyAllowance={1400}
        daysElapsed={10}
        daysRemaining={22}
        intervalStartDate={25}
        totalAllowance={43400}
        totalSpent={13259}
        updateAllowance={() => null}
        updateIntervalStartDate={() => null}
      />
    );
    const debitAmount = output.findWhere(
      node => node.prop("testID") === "averageRemain"
    );
    expect(debitAmount.first().prop("children")).toEqual("13.70");
    const output2 = shallow(
      <RemainderCalculations
        dailyAllowance={1000}
        daysElapsed={31}
        daysRemaining={1}
        intervalStartDate={4}
        totalAllowance={31000}
        totalSpent={30000}
        updateAllowance={() => null}
        updateIntervalStartDate={() => null}
      />
    );
    const debitAmount2 = output2.findWhere(
      node => node.prop("testID") === "averageRemain"
    );
    expect(debitAmount2.first().prop("children")).toEqual("10.00");
  });

  it("calculates the total amount remaining", () => {
    const output = shallow(
      <RemainderCalculations
        dailyAllowance={1000}
        daysElapsed={10}
        daysRemaining={22}
        intervalStartDate={25}
        totalAllowance={31000}
        totalSpent={21000}
        updateAllowance={() => null}
        updateIntervalStartDate={() => null}
      />
    );
    const amountRemaining = output.findWhere(
      node => node.prop("testID") === "totalRemain"
    );
    expect(amountRemaining.first().prop("children")).toEqual("100.00");
  });
});
