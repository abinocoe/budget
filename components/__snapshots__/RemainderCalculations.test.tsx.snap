// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemainderCalculations renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "justifyContent": "space-evenly",
      "marginTop": 20,
    }
  }
>
  <View
    style={
      Object {
        "justifyContent": "flex-start",
      }
    }
  >
    <PeriodAllowanceInput
      allowanceAmount={30000}
      updateAllowance={[Function]}
    />
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Array [
            Object {
              "borderRadius": 5,
              "color": "white",
              "elevation": 3,
              "fontSize": 30,
              "marginBottom": 5,
              "marginTop": 10,
              "paddingHorizontal": 5,
              "textAlign": "center",
              "width": "90%",
            },
            Object {
              "backgroundColor": "#AB47BC",
            },
          ]
        }
        testID="totalRemain"
      >
        270.00
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Total remaining
      </Text>
    </View>
    <DateSelection
      intervalStartDate={25}
      updateIntervalStartDate={[Function]}
    />
  </View>
  <View
    style={
      Object {
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Array [
            Object {
              "borderRadius": 5,
              "color": "white",
              "elevation": 3,
              "fontSize": 30,
              "marginBottom": 5,
              "marginTop": 10,
              "paddingHorizontal": 5,
              "textAlign": "center",
              "width": "90%",
            },
            Object {
              "backgroundColor": "#4CAF50",
            },
          ]
        }
      >
        10.00
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Daily allowance
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Array [
            Object {
              "borderRadius": 5,
              "color": "white",
              "elevation": 3,
              "fontSize": 30,
              "marginBottom": 5,
              "marginTop": 10,
              "paddingHorizontal": 5,
              "textAlign": "center",
              "width": "90%",
            },
            Object {
              "backgroundColor": "#FFEB3B",
            },
          ]
        }
        testID="averageRemain"
      >
        9.31
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Average per day remain
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
        }
      }
    >
      <Text
        onPress={[Function]}
        style={
          Array [
            Object {
              "borderRadius": 5,
              "color": "white",
              "elevation": 3,
              "fontSize": 30,
              "marginBottom": 5,
              "marginTop": 10,
              "paddingHorizontal": 5,
              "textAlign": "center",
              "width": "90%",
            },
            Object {
              "backgroundColor": "#FF9800",
            },
          ]
        }
        testID="amountDebit"
      >
        -10.00
      </Text>
      <Text
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Amount debit
      </Text>
    </View>
  </View>
  <ReactNativeModal
    animationIn="slideInUp"
    animationInTiming={300}
    animationOut="slideOutDown"
    animationOutTiming={300}
    avoidKeyboard={false}
    backdropColor="black"
    backdropOpacity={0.7}
    backdropTransitionInTiming={300}
    backdropTransitionOutTiming={300}
    deviceHeight={null}
    deviceWidth={null}
    hasBackdrop={true}
    hideModalContentWhileAnimating={false}
    isVisible={false}
    onBackButtonPress={[Function]}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalShow={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    useNativeDriver={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "justifyContent": "center",
        }
      }
    >
      <Text />
    </View>
  </ReactNativeModal>
</View>
`;
